substrate/client/rpc-spec-v2/src/transaction/transaction_broadcast.rs

```diff
@@ -0,0 +1,217 @@
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! API implementation for broadcasting transactions.
+
+use crate::{transaction::api::TransactionBroadcastApiServer, SubscriptionTaskExecutor};
+use codec::Decode;
+use futures::{FutureExt, StreamExt};
+use futures_util::stream::AbortHandle;
+use jsonrpsee::core::{async_trait, RpcResult};
+use parking_lot::RwLock;
+use rand::{distributions::Alphanumeric, Rng};
+use sc_client_api::BlockchainEvents;
+use sc_transaction_pool_api::{
+	BlockHash, TransactionFor, TransactionPool, TransactionSource, TransactionStatus,
+};
+use sp_blockchain::HeaderBackend;
+use sp_core::Bytes;
+use sp_runtime::traits::Block as BlockT;
+use std::{collections::HashMap, sync::Arc};
+
+use super::error::ErrorBroadcast;
+
+/// An API for transaction RPC calls.
+pub struct TransactionBroadcast<Pool, Client> {
+	/// Substrate client.
+	client: Arc<Client>,
+	/// Transactions pool.
+	pool: Arc<Pool>,
+	/// Executor to spawn subscriptions.
+	executor: SubscriptionTaskExecutor,
+	/// The brodcast operation IDs.
+	broadcast_ids: Arc<RwLock<HashMap<String, BroadcastState>>>,
+}
+
+/// The state of a broadcast operation.
+struct BroadcastState {
+	/// Handle to abort the running future that broadcasts the transaction.
+	handle: AbortHandle,
+}
+
+impl<Pool, Client> TransactionBroadcast<Pool, Client> {
+	/// Creates a new [`TransactionBroadcast`].
+	pub fn new(client: Arc<Client>, pool: Arc<Pool>, executor: SubscriptionTaskExecutor) -> Self {
+		TransactionBroadcast { client, pool, executor, broadcast_ids: Default::default() }
+	}
+
+	/// Generate an unique operation ID for the `transaction_broadcast` RPC method.
+	pub fn generate_unique_id(&self) -> String {
+		let generate_operation_id = || {
+			// The lenght of the operation ID.
+			const OPERATION_ID_LEN: usize = 16;
+
+			rand::thread_rng()
+				.sample_iter(Alphanumeric)
+				.take(OPERATION_ID_LEN)
+				.map(char::from)
+				.collect::<String>()
+		};
+
+		let mut id = generate_operation_id();
+
+		let broadcast_ids = self.broadcast_ids.read();
+
+		while broadcast_ids.contains_key(&id) {
+			id = generate_operation_id();
+		}
+
+		id
+	}
+}
+
+/// Currently we treat all RPC transactions as externals.
+///
+/// Possibly in the future we could allow opt-in for special treatment
+/// of such transactions, so that the block authors can inject
+/// some unique transactions via RPC and have them included in the pool.
+const TX_SOURCE: TransactionSource = TransactionSource::External;
+
+#[async_trait]
+impl<Pool, Client> TransactionBroadcastApiServer<BlockHash<Pool>>
+	for TransactionBroadcast<Pool, Client>
+where
+	Pool: TransactionPool + Sync + Send + 'static,
+	Pool::Hash: Unpin,
+	<Pool::Block as BlockT>::Hash: Unpin,
+	Client: HeaderBackend<Pool::Block> + BlockchainEvents<Pool::Block> + Send + Sync + 'static,
+{
+	fn broadcast(&self, bytes: Bytes) -> RpcResult<Option<String>> {
```

Comment:
why is the return type `Option<String>` it looks like always an ID is returned?!